src/app/clusters/tls-certificate-management-server/tls-certificate-management-server.cpp

```diff
@@ -0,0 +1,443 @@
+                                                            return CHIP_NO_ERROR;
+                                                        });
+    if (result == CHIP_ERROR_NOT_FOUND)
+    {
+        ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::NotFound);
+    }
+    else if (result != CHIP_NO_ERROR)
+    {
+        ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::Failure);
+    }
+}
+
+void TlsCertificateManagementServer::HandleRemoveRootCertificate(HandlerContext & ctx,
+                                                                 const RemoveRootCertificate::DecodableType & req)
+{
+    ChipLogDetail(Zcl, "TlsCertificateManagement: RemoveRootCertificate");
+    auto result = mDelegate.RemoveRootCert(ctx.mRequestPath.mEndpointId, ctx.mCommandHandler.GetAccessingFabricIndex(), req.caid);
+    if (result == CHIP_ERROR_NOT_FOUND)
+    {
+        ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::NotFound);
+    }
+    else if (result == CHIP_NO_ERROR)
+    {
+        ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::Success);
+    }
+    else
+    {
+        ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::Failure);
+    }
+}
+
+void TlsCertificateManagementServer::HandleGenerateClientCsr(HandlerContext & ctx, const TLSClientCSR::DecodableType & req)
+{
+    ChipLogDetail(Zcl, "TlsCertificateManagement: TLSClientCSR");
+
+    VerifyOrReturn(req.nonce.size() <= 128, ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::ConstraintError));
+
+    auto status = mDelegate.GenerateClientCsr(
+        ctx.mRequestPath.mEndpointId, ctx.mCommandHandler.GetAccessingFabricIndex(), req, [&](auto & response) -> Status {
+            VerifyOrDieWithMsg(response.ccdid <= kMaxClientCertId, NotSpecified, "Spec requires CCDID to be <= kMaxClientCertId");
+            VerifyOrDieWithMsg(response.csr.size() <= 3000, NotSpecified, "Spec requires csr.size() to be <= 3000");
+            VerifyOrDieWithMsg(response.nonce.size() <= 128, NotSpecified, "Spec requires nonce.size() to be < kMaxClientCertId");
+            ctx.mCommandHandler.AddResponse(ctx.mRequestPath, response);
+            return Status::Success;
+        });
+
+    if (status != Status::Success)
+    {
+        ctx.mCommandHandler.AddStatus(ctx.mRequestPath, status);
+    }
+}
+
+void TlsCertificateManagementServer::HandleProvisionClientCertificate(HandlerContext & ctx,
+                                                                      const ProvisionClientCertificate::DecodableType & req)
+{
+    ChipLogDetail(Zcl, "TlsCertificateManagement: ProvisionClientCertificate");
+
+    VerifyOrReturn(req.ccdid <= kMaxClientCertId, ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::ConstraintError));
+    const auto & detail = req.clientCertificateDetails;
+    VerifyOrReturn(!detail.clientCertificate.HasValue() || detail.clientCertificate.Value().size() <= kSpecMaxCertBytes,
```

Comment:
I don't understand this part.  In the spec ClientCertificate is neither optional nor nullable.... and there is no ClientCertificateDetails at all?

This really doesn't match what I see in the spec right now.

I'm going to stop here, in HandleProvisionClientCertificate, because it's clear this is implementing an older version of the spec, which makes it a bit hard to review...  Either point me to the version being implemented, please, or update the XML to the spec and then update this code?